designs/disruption-controls.md

```diff
@@ -0,0 +1,118 @@
+# Disruption Controls - (in v1beta1+)
+
+## Motivation
+Users have been asking for more control around the speed of Consolidation and the ability to restrict disruptions to their nodes. The demand for this is reflected by the +1s in [#1738](https://github.com/aws/karpenter/issues/1738). This is increasingly important as clusters can run at a highly varied scales, where some users have reported scale-down behaving too slow at higher scales, or too quick at smaller scales. Additionally, users that want to control when nodes can be disrupted must add PDBs to block node disruptions, add `do-not-evict` annotations on pods for sensitive workloads, or add `do-not-consolidate` annotations on nodes that shouldn't be consolidated. Natively integrating a better way to ensure nodes aren't disrupted will ease the burden that users have to take on with disruption.
+
+These controls will include a new `Disruption` block in the `v1beta1` `NodePool` that will contain all scale-down behavioral fields. Initially, this will include `Budgets`, which define a (1) parallelism of how many nodes can be deprovisioned at a time and a (2) cron schedule that determines when the parallelism applies, and a `ConsolidateAfter` field that will allow users to affect the speed at which Karpenter will scale down underutilized nodes. Future scale-down behavioral fields should be colocated with these fields.
+
+## Proposed Spec
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized || WhenEmpty
+    consolidateAfter: 10m || Never # metav1.Duration
+    expireAfter: 10m || Never # Equivalent to v1alpha5 TTLSecondsUntilExpired
+    budgets:
+    # On Weekdays during business hours, don't do any deprovisioning.
+    - crontab: "0 9 * * mon-fri"
+      duration: 8h
+      maxUnavailable: 0
+    # Every other time, only allow 10 nodes to be deprovisioned simultaneously
+    - maxUnavailable: 10
```

Comment:
Can a schedule like this optionally force consolidation/drift on nodes with `do-not-evict` pods?

Reply:
There's a some issues in karpenter/#1738 that are intentionally not included in this design doc that would probably handle force-delete cases with `do-not-evict` pods. This Budget is meant to define periods where Karpenter cannot disrupt, rather than periods where Karpenter can more forcefully disrupt. 

Can you share some info about your use-case here?

Reply:
We're effectively trying to automate all normal cluster maintenance like node updates, TTLs, etc but a lot of our customers don't want this triggering at any moment. 

One consolation we've made is that we can do whatever during normal scheduled periods similar to RDS maintenance windows. 

So right now the biggest issues for us is that there is no ability to control drift beyond enable disable globally and the do not evict annotations have no way to be dealt with automatically. Our 2 primary users of the annotation are batch type jobs which will finish so are less of an issue and long running singletons that don't want/can't deal with disruption.

Reply:
I think https://github.com/aws/karpenter/issues/1526 would be the best for you here, as you ensure Karpenter doesn't disrupt these important applications, but can ensure that this do-not-disrupt pod will be evicted once it's reached an expected grace period.

Generally, we want drift to be the default experience on how Karpenter works, but enable users to determine when they want their nodes disrupted through enough pod/node annotations or conditions. 

More granular pod/node level controls is something we're looking to tackle as a second round of disruption controls after this first round.